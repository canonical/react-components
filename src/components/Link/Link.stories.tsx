import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import Link from "./Link";
import Strip from "../Strip";

const meta: Meta<typeof Link> = {
  component: Link,
  tags: ["autodocs"],

  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },
  },
};

export default meta;

type Story = StoryObj<typeof Link>;

export const Default: Story = {
  name: "Default",

  args: {
    children:
      "Fast, secure and simple, Ubuntu powers millions of PCs worldwide",
  },
};

export const Soft: Story = {
  render: () => <Link soft>Learn about MAAS</Link>,
  name: "Soft",
};

export const Inverted: Story = {
  render: () => (
    <Strip type="dark">
      <Link inverted>Package & publish your app</Link>
    </Strip>
  ),

  name: "Inverted",
};

export const BackToTop: Story = {
  render: () => <Link top>Back to top</Link>,
  name: "Back to top",
};
