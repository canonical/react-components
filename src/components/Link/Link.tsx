import classNames from "classnames";
import React from "react";
import type { HTMLProps, ReactNode } from "react";

import type { ClassName, PropsWithSpread } from "types";

export type Props = PropsWithSpread<
  {
    /**
     * The content of the link.
     */
    children: ReactNode;
    /**
     * Optional class(es) to pass to the wrapping anchor element.
     */
    className?: ClassName;
    /**
     * Whether the link should have inverted styling.
     */
    inverted?: boolean;
    /**
     * Whether the link should have soft styling.
     */
    soft?: boolean;
    /**
     * Whether the link should have "back to top" styling.
     */
    top?: boolean;
  },
  HTMLProps<HTMLAnchorElement>
>;

/**
 * This is a [React](https://reactjs.org/) component for the Vanilla [Link](https://docs.vanillaframework.io/patterns/links/).
 *
 * Links are used to embed actions or pathways to more information, allowing users to click their way from page to page.
 */
const Link = ({
  children,
  className,
  href = "#",
  inverted = false,
  soft = false,
  top = false,
  ...props
}: Props): JSX.Element => {
  const link = (
    <a
      className={classNames(className, {
        "p-link--inverted": inverted,
        "p-link--soft": soft,
        "p-top__link": top,
      })}
      href={href}
      {...props}
    >
      {children}
    </a>
  );
  if (top) {
    return <div className="p-top">{link}</div>;
  }
  return link;
};

export default Link;
