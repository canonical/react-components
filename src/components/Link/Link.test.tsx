import { render, screen } from "@testing-library/react";
import React from "react";

import Link from "./Link";

describe("Link ", () => {
  it("renders", () => {
    render(<Link href="/example">Test content</Link>);
    expect(screen.getByRole("link")).toMatchSnapshot();
  });

  it("can be an inverted link", () => {
    render(<Link inverted={true}>Test content</Link>);
    expect(screen.getByRole("link")).toHaveClass("p-link--inverted");
  });

  it("can be a soft link", () => {
    render(<Link soft={true}>Test content</Link>);
    expect(screen.getByRole("link")).toHaveClass("p-link--soft");
  });

  it("can be a back to top link", () => {
    render(<Link top={true}>Back to top</Link>);
    expect(screen.getByRole("link")).toMatchSnapshot();
  });

  it("can add additional classes", () => {
    render(
      <Link soft={true} className="extra-class">
        Test content
      </Link>
    );
    const link = screen.getByRole("link");
    expect(link).toHaveClass("p-link--soft");
    expect(link).toHaveClass("extra-class");
  });
});
