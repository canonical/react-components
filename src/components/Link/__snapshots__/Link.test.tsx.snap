// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link  can be a back to top link 1`] = `
<a
  class="p-top__link"
  href="#"
>
  Back to top
</a>
`;

exports[`Link  renders 1`] = `
<a
  class=""
  href="/example"
>
  Test content
</a>
`;
